ocaml/xapi/xapi_xenops.ml

```diff
@@ -1039,29 +1039,41 @@ module Xapi_cache = struct
       		updated whenever we receive an event from xapi. *)
 
   let cache = Hashtbl.create 10 (* indexed by Vm.id *)
+  let mutex = Mutex.create ()
+  let with_lock f =
+    Mutex.execute mutex f
+
+  let register id initial_value =
+    with_lock (fun () ->
+        debug "xapi_cache: creating cache for %s" id;
+        if not(Hashtbl.mem cache id) || (Hashtbl.find cache id = None)
```

Comment:
```
if (try Hashtbl.find cache id = None with Not_found -> true) then
```

Probing the hash table twice is an anti pattern. Because of the lock we don't get a race here but I think it should be still avoided.